singer/src/main/java/com/pinterest/singer/SingerMain.java

```diff
@@ -54,6 +54,14 @@ public void run() {
       } catch (Throwable t) {
         LOG.error("Shutdown failure: log monitor : ", t);
       }
+      
```

Comment:
Instead of introducing another daemon thread,  how about using ostrich to collect kafka producer metrics like all other metrics that are being collected now?

Reply:
There are two downsides to that approach.

1. Metrics are conflicting and duplicate as a single producer may be shared by multiple kafkawriter instances
2. Lack of centralized management of producer level metrics

An independent thread makes for a clean and non-intrusive design.